File: lib/ocaml/test/test_recursion_depth.ml
(*
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements. See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership. The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License. You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied. See the License for the
 specific language governing permissions and limitations
 under the License.
*)

(* Recursion-depth guard test for the OCaml library (THRIFT-6051).

   This exercises the real depth guard that lives in [Thrift.Protocol.t]
   (increment_recursion_depth / decrement_recursion_depth, limit 64) through
   full struct/union/exception read+write round-trips, NOT by calling those
   two methods in isolation.

   The recursive types and their read/write functions below are written by
   hand to mirror exactly what the OCaml generator now emits (the same
   "increment_recursion_depth; Fun.protect ~finally:decrement (fun () -> ...)"
   wrapper around every struct reader and writer). Generated code from
   test/Recursive.thrift cannot be used directly here for two pre-existing,
   unrelated reasons: the generator emits mutually recursive classes
   (CoRec/CoUnion/CoError pairs) as separate "class" declarations instead of
   "class ... and ...", and TBinaryProtocol.ml does not compile on modern
   OCaml (uses removed mutable-string APIs). Both are out of scope for
   THRIFT-6051 and flagged separately.

   In the OCaml generator unions and exceptions are emitted through the very
   same code path as structs (generate_ocaml_struct_definition), so a single
   guarded reader/writer template covers all three. The three node types below
   make that coverage explicit at runtime. *)

open Thrift

(* Minimal no-op transport: the in-memory protocol below never touches it. *)
class null_transport =
object
  inherit Transport.t
  method isOpen = true
  method opn = ()
  method close = ()
  method read _buf _off _len = 0
  method write _buf _off _len = ()
  method flush = ()
end

(* A tiny in-memory protocol that records the structural events as tokens in a
   FIFO queue on write and replays them on read. It implements only the methods
   the recursive types below use; the rest fail loudly if ever reached. This
   stands in for TBinaryProtocol, which does not compile on modern OCaml. *)
type token =
  | StructBegin of string
  | StructEnd
  | FieldBegin of string * Protocol.t_type * int
  | FieldEnd
  | FieldStop
  | I16 of int

class mem_protocol trans =
object (self)
  inherit Protocol.t trans

  val tokens : token Queue.t = Queue.create ()
  method private put t = Queue.add t tokens
  method private get = Queue.take tokens

  (* functional, structural methods *)
  method writeStructBegin name = self#put (StructBegin name)
  method writeStructEnd = self#put StructEnd
  method writeFieldBegin (name, t, id) = self#put (FieldBegin (name, t, id))
  method writeFieldEnd = self#put FieldEnd
  method writeFieldStop = self#put FieldStop
  method writeI16 i = self#put (I16 i)

  method readStructBegin =
    (match self#get with StructBegin n -> n | _ -> failwith "expected StructBegin")
  method readStructEnd =
    (match self#get with StructEnd -> () | _ -> failwith "expected StructEnd")
  method readFieldBegin =
    (match self#get with
     | FieldBegin (n, t, id) -> (n, t, id)
     | FieldStop -> ("", Protocol.T_STOP, 0)
     | _ -> failwith "expected FieldBegin or FieldStop")
  method readFieldEnd =
    (match self#get with FieldEnd -> () | _ -> failwith "expected FieldEnd")
  method readI16 =
    (match self#get with I16 i -> i | _ -> failwith "expected I16")

  (* unused by this test *)
  method writeMessageBegin _ = failwith "unused"
  method writeMessageEnd = failwith "unused"
  method writeMapBegin _ = failwith "unused"
  method writeMapEnd = failwith "unused"
  method writeListBegin _ = failwith "unused"
  method writeListEnd = failwith "unused"
  method writeSetBegin _ = failwith "unused"
  method writeSetEnd = failwith "unused"
  method writeBool _ = failwith "unused"
  method writeByte _ = failwith "unused"
  method writeI32 _ = failwith "unused"
  method writeI64 _ = failwith "unused"
  method writeDouble _ = failwith "unused"
  method writeString _ = failwith "unused"
  method writeBinary _ = failwith "unused"
  method readMessageBegin = failwith "unused"
  method readMessageEnd = failwith "unused"
  method readMapBegin = failwith "unused"
  method readMapEnd = failwith "unused"
  method readListBegin = failwith "unused"
  method readListEnd = failwith "unused"
  method readSetBegin = failwith "unused"
  method readSetEnd = failwith "unused"
  method readBool = failwith "unused"
  method readByte = failwith "unused"
  method readI32 = failwith "unused"
  method readI64 = failwith "unused"
  method readDouble = failwith "unused"
  method readString = failwith "unused"
  method readBinary = failwith "unused"
end

let passed = ref 0
let failed = ref 0

let check label cond =
  if cond then begin
    Printf.printf "PASS: %s\n%!" label;
    incr passed
  end else begin
    Printf.printf "FAIL: %s\n%!" label;
    incr failed
  end

(* ---- Recursive STRUCT (self-recursive, mirrors generated struct codegen) ---- *)
class rec_struct =
object (_self)
  val mutable _other : rec_struct option = None
  val mutable _item : int = 0
  method get_other = _other
  method set_other (x : rec_struct) = _other <- Some x
  method set_item i = _item <- i
  method write (oprot : Protocol.t) =
    oprot#increment_recursion_depth;
    Fun.protect ~finally:(fun () -> oprot#decrement_recursion_depth) (fun () ->
      oprot#writeStructBegin "RecStruct";
      (match _other with None -> () | Some _v ->
        oprot#writeFieldBegin ("other", Protocol.T_STRUCT, 1);
        _v#write oprot;
        oprot#writeFieldEnd);
      oprot#writeFieldBegin ("item", Protocol.T_I16, 2);
      oprot#writeI16 _item;
      oprot#writeFieldEnd;
      oprot#writeFieldStop;
      oprot#writeStructEnd)
end

let rec read_rec_struct (iprot : Protocol.t) =
  let str = new rec_struct in
  iprot#increment_recursion_depth;
  Fun.protect ~finally:(fun () -> iprot#decrement_recursion_depth) (fun () ->
    ignore (iprot#readStructBegin);
    (try while true do
      let (_, t, id) = iprot#readFieldBegin in
      if t = Protocol.T_STOP then raise Break;
      (match id with
       | 1 -> str#set_other (read_rec_struct iprot)
       | 2 -> str#set_item iprot#readI16
       | _ -> iprot#skip t);
      iprot#readFieldEnd
    done with Break -> ());
    iprot#readStructEnd);
  str

let rec build_struct n =
  let s = new rec_struct in
  s#set_item n;
  if n > 1 then s#set_other (build_struct (n - 1));
  s

let rec depth_struct s =
  match s#get_other with None -> 1 | Some o -> 1 + depth_struct o

(* Craft an over-limit payload with raw (unguarded) protocol calls, so the
   depth limit can only trip on the read side. *)
let rec craft_struct (oprot : Protocol.t) n =
  oprot#writeStructBegin "RecStruct";
  if n > 1 then begin
    oprot#writeFieldBegin ("other", Protocol.T_STRUCT, 1);
    craft_struct oprot (n - 1);
    oprot#writeFieldEnd
  end;
  oprot#writeFieldStop;
  oprot#writeStructEnd

(* ---- Recursive UNION (OCaml emits unions through the struct path) ---- *)
class rec_union =
object (_self)
  val mutable _other : rec_union option = None
  method get_other = _other
  method set_other (x : rec_union) = _other <- Some x
  method write (oprot : Protocol.t) =
    oprot#increment_recursion_depth;
    Fun.protect ~finally:(fun () -> oprot#decrement_recursion_depth) (fun () ->
      oprot#writeStructBegin "RecUnion";
      (match _other with None -> () | Some _v ->
        oprot#writeFieldBegin ("other", Protocol.T_STRUCT, 1);
        _v#write oprot;
        oprot#writeFieldEnd);
      oprot#writeFieldStop;
      oprot#writeStructEnd)
end

let rec read_rec_union (iprot : Protocol.t) =
  let str = new rec_union in
  iprot#increment_recursion_depth;
  Fun.protect ~finally:(fun () -> iprot#decrement_recursion_depth) (fun () ->
    ignore (iprot#readStructBegin);
    (try while true do
      let (_, t, id) = iprot#readFieldBegin in
      if t = Protocol.T_STOP then raise Break;
      (match id with
       | 1 -> str#set_other (read_rec_union iprot)
       | _ -> iprot#skip t);
      iprot#readFieldEnd
    done with Break -> ());
    iprot#readStructEnd);
  str

let rec build_union n =
  let u = new rec_union in
  if n > 1 then u#set_other (build_union (n - 1));
  u

let rec depth_union u =
  match u#get_other with None -> 1 | Some o -> 1 + depth_union o

let rec craft_union (oprot : Protocol.t) n =
  oprot#writeStructBegin "RecUnion";
  if n > 1 then begin
    oprot#writeFieldBegin ("other", Protocol.T_STRUCT, 1);
    craft_union oprot (n - 1);
    oprot#writeFieldEnd
  end;
  oprot#writeFieldStop;
  oprot#writeStructEnd

(* ---- Recursive EXCEPTION (OCaml emits a class plus an "exception E of cls") ---- *)
class rec_error =
object (_self)
  val mutable _other : rec_error option = None
  method get_other = _other
  method set_other (x : rec_error) = _other <- Some x
  method write (oprot : Protocol.t) =
    oprot#increment_recursion_depth;
    Fun.protect ~finally:(fun () -> oprot#decrement_recursion_depth) (fun () ->
      oprot#writeStructBegin "RecError";
      (match _other with None -> () | Some _v ->
        oprot#writeFieldBegin ("other", Protocol.T_STRUCT, 1);
        _v#write oprot;
        oprot#writeFieldEnd);
      oprot#writeFieldStop;
      oprot#writeStructEnd)
end
exception RecError of rec_error

let rec read_rec_error (iprot : Protocol.t) =
  let str = new rec_error in
  iprot#increment_recursion_depth;
  Fun.protect ~finally:(fun () -> iprot#decrement_recursion_depth) (fun () ->
    ignore (iprot#readStructBegin);
    (try while true do
      let (_, t, id) = iprot#readFieldBegin in
      if t = Protocol.T_STOP then raise Break;
      (match id with
       | 1 -> str#set_other (read_rec_error iprot)
       | _ -> iprot#skip t);
      iprot#readFieldEnd
    done with Break -> ());
    iprot#readStructEnd);
  str

let rec build_error n =
  let e = new rec_error in
  if n > 1 then e#set_other (build_error (n - 1));
  e

let rec depth_error e =
  match e#get_other with None -> 1 | Some o -> 1 + depth_error o

let rec craft_error (oprot : Protocol.t) n =
  oprot#writeStructBegin "RecError";
  if n > 1 then begin
    oprot#writeFieldBegin ("other", Protocol.T_STRUCT, 1);
    craft_error oprot (n - 1);
    oprot#writeFieldEnd
  end;
  oprot#writeFieldStop;
  oprot#writeStructEnd

let limit = 64

let is_depth_limit = function Protocol.E (Protocol.DEPTH_LIMIT, _) -> true | _ -> false

(* Run the full read/write round-trip matrix for one node category. *)
let run_suite label ~build ~write ~read ~depth_of ~craft =
  (* 1. A chain exactly at the limit round-trips and preserves its depth. *)
  (let proto = new mem_protocol (new null_transport) in
   try
     write (build limit) proto;
     let got = depth_of (read proto) in
     check (Printf.sprintf "%s: %d-deep write/read round-trip preserves depth" label limit)
       (got = limit)
   with e ->
     Printf.printf "  (unexpected: %s)\n%!" (Printexc.to_string e);
     check (Printf.sprintf "%s: %d-deep write/read round-trip preserves depth" label limit) false);

  (* 2. Writing one level past the limit raises DEPTH_LIMIT. *)
  (let proto = new mem_protocol (new null_transport) in
   try
     write (build (limit + 1)) proto;
     check (Printf.sprintf "%s: write %d-deep raises DEPTH_LIMIT" label (limit + 1)) false
   with e ->
     check (Printf.sprintf "%s: write %d-deep raises DEPTH_LIMIT" label (limit + 1))
       (is_depth_limit e));

  (* 3. Reading a crafted over-limit payload raises DEPTH_LIMIT. *)
  (let proto = new mem_protocol (new null_transport) in
   craft proto (limit + 1);
   try
     ignore (read proto);
     check (Printf.sprintf "%s: read %d-deep raises DEPTH_LIMIT" label (limit + 1)) false
   with e ->
     check (Printf.sprintf "%s: read %d-deep raises DEPTH_LIMIT" label (limit + 1))
       (is_depth_limit e))

let () =
  run_suite "struct"
    ~build:build_struct
    ~write:(fun o p -> o#write p)
    ~read:read_rec_struct
    ~depth_of:depth_struct
    ~craft:craft_struct;
  run_suite "union"
    ~build:build_union
    ~write:(fun o p -> o#write p)
    ~read:read_rec_union
    ~depth_of:depth_union
    ~craft:craft_union;
  run_suite "exception"
    ~build:build_error
    ~write:(fun o p -> o#write p)
    ~read:read_rec_error
    ~depth_of:depth_error
    ~craft:craft_error;

  (* The exception wrapper itself is exercised so the generated
     "exception E of cls" shape is covered as well. *)
  (try raise (RecError (build_error 1)) with RecError _ -> check "exception wrapper raises/catches" true);

  Printf.printf "\nResults: %d passed, %d failed\n" !passed !failed;
  if !failed > 0 then exit 1

File: lib/lua/test/test_recursion_depth.lua
--
-- Licensed to the Apache Software Foundation (ASF) under one
-- or more contributor license agreements. See the NOTICE file
-- distributed with this work for additional information
-- regarding copyright ownership. The ASF licenses this file
-- to you under the Apache License, Version 2.0 (the
-- "License"); you may not use this file except in compliance
-- with the License. You may obtain a copy of the License at
--
--   http://www.apache.org/licenses/LICENSE-2.0
--
-- Unless required by applicable law or agreed to in writing,
-- software distributed under the License is distributed on an
-- "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
-- KIND, either express or implied. See the License for the
-- specific language governing permissions and limitations
-- under the License.
--

-- Drives the recursion-depth guard through the *generated* read()/write()
-- code over a recursive struct (RecTree), union (RecUnion) and exception
-- (RecError), not the protocol counter in isolation.
--
-- Run after generating the types next to this file:
--   thrift -o lib/lua/test --gen lua lib/lua/test/RecursionDepth.thrift
--   lua lib/lua/test/test_recursion_depth.lua

-- Locate the library (../) and the generated code (gen-lua/) relative to this
-- script so it can be run from anywhere.
local script_dir = arg[0]:match('(.*[/\\])') or './'
package.path = script_dir .. '../?.lua;' ..
               script_dir .. 'gen-lua/?.lua;' .. package.path

-- The recursion guard under test is pure Lua (the generated read/write and the
-- TProtocolBase counter). The C extensions only do byte packing, which Lua 5.3+
-- provides natively, so functional pure-Lua stand-ins let the test run without
-- building them while still exercising the real generated serialization path.
package.preload['libluabitwise'] = function()
  return {
    bor    = function(a, b) return a | b end,
    band   = function(a, b) return a & b end,
    bxor   = function(a, b) return a ~ b end,
    shiftl = function(a, n) return (a << n) & 0xFFFFFFFF end,
    shiftr = function(a, n) return a >> n end,
  }
end
package.preload['libluabpack'] = function()
  local fmt = {c = '>i1', C = '>I1', s = '>i2', S = '>I2',
               i = '>i4', I = '>I4', l = '>i8', d = '>d'}
  return {
    bpack   = function(code, val) return string.pack(fmt[code], val) end,
    bunpack = function(code, data) return (string.unpack(fmt[code], data)) end,
  }
end
package.preload['liblualongnumber'] = function()
  -- i64 is never exercised here (the test types use i16/i32 only).
  return {new = function(_, v) return v or 0 end, tonumber = function(v) return v end}
end

require('Thrift')
require('TTransport')
require('TMemoryBuffer')
require('TProtocol')
require('TBinaryProtocol')
require('RecursionDepth_ttypes')

local LIMIT = DEFAULT_RECURSION_DEPTH

local passed, failed = 0, 0
local function ok(cond, name, detail)
  if cond then
    print('ok - ' .. name)
    passed = passed + 1
  else
    print('not ok - ' .. name .. (detail and ('  # ' .. tostring(detail)) or ''))
    failed = failed + 1
  end
end

local function new_proto()
  return TBinaryProtocol:new{trans = TMemoryBuffer:new{}}
end

-- Build a chain of `depth` nested nodes; each inner node carries a single
-- child, the deepest carries the scalar leaf -- a valid shape for both the
-- struct and the union.
local function make_chain(class, leaf_field, depth)
  local node = class:new{}
  if depth > 1 then
    node.children = {make_chain(class, leaf_field, depth - 1)}
  else
    node[leaf_field] = 1
  end
  return node
end

local function chain_depth(node)
  local depth = 1
  if node.children and #node.children > 0 then
    depth = depth + chain_depth(node.children[1])
  end
  return depth
end

-- Serialize an over-limit payload with raw protocol primitives so the reader
-- recurses through the guarded struct path (field id 1 = list<self>), not the
-- separate (unbounded) skip() path.
local function write_deep(oprot, depth)
  oprot:writeStructBegin('Rec')
  if depth > 1 then
    oprot:writeFieldBegin('children', TType.LIST, 1)
    oprot:writeListBegin(TType.STRUCT, 1)
    write_deep(oprot, depth - 1)
    oprot:writeListEnd()
    oprot:writeFieldEnd()
  end
  oprot:writeFieldStop()
  oprot:writeStructEnd()
end

-- True only for the depth-limit rejection -- an EOF/structural error would
-- carry a different type/message.
local function is_depth_error(err)
  return type(err) == 'string'
     and err:find('TProtocolException') ~= nil
     and err:find('recursion') ~= nil
end

local cases = {
  {kind = 'struct',    class = RecTree,  leaf = 'item'},
  {kind = 'union',     class = RecUnion, leaf = 'leaf'},
  {kind = 'exception', class = RecError, leaf = 'leaf'},
}

for _, case in ipairs(cases) do
  local kind, class, leaf = case.kind, case.class, case.leaf

  -- 1. A chain exactly at the limit round-trips. Also proves the guard does
  --    not double-count (a chain of 64 would be rejected at 32 otherwise).
  do
    local proto = new_proto()
    local chain = make_chain(class, leaf, LIMIT)
    local wok, werr = pcall(function() chain:write(proto) end)
    ok(wok, kind .. ': writing a chain at the depth limit succeeds', werr)

    local decoded = class:new{}
    local rok, rerr = pcall(function() decoded:read(proto) end)
    ok(rok, kind .. ': reading a chain at the depth limit succeeds', rerr)
    ok(chain_depth(decoded) == LIMIT,
       kind .. ': round-trips to the original depth (' .. LIMIT .. ')')
  end

  -- 2. Writing past the limit is rejected with a depth-limit error.
  do
    local proto = new_proto()
    local chain = make_chain(class, leaf, LIMIT + 5)
    local wok, werr = pcall(function() chain:write(proto) end)
    ok(not wok, kind .. ': writing past the limit throws')
    ok(is_depth_error(werr), kind .. ': ... with a recursion-depth error', werr)
  end

  -- 3. Reading an over-limit payload is rejected with a depth-limit error.
  do
    local proto = new_proto()
    write_deep(proto, LIMIT + 5)
    local decoded = class:new{}
    local rok, rerr = pcall(function() decoded:read(proto) end)
    ok(not rok, kind .. ': reading past the limit throws')
    ok(is_depth_error(rerr), kind .. ': ... with a recursion-depth error', rerr)
  end
end

print(string.format('\n%d passed, %d failed', passed, failed))
if failed > 0 then os.exit(1) end
